typing/mode.ml

```diff
@@ -2253,9 +2253,20 @@ module Modality = struct
         | Join_const c -> Format.fprintf ppf "join_const(%a)" Mode.Const.print c
     end
 
+    (* Similar to constant modalities, an inferred modality maps the mode of a
+       record/structure to the mode of a value therein. An inferred modality [f] is
+       inferred from the structure/record mode [mm] and the value mode [m].
+
+       Soundness: You should not get a value from a record/structure at a mode strictly
+       stronger than how it was put in. That is, [f mm >= m].
```

Comment:
Is this right? It seems OK for comonadic axes, but I don’t buy it for monadic ones.

Reply:
This is true for both axes (what you get out can't be stronger than what you put in).

Additionally, there is another soundness condition: for comonadic axes, upon construction the value must be stronger than the module. For monadic axes, upon destruction the value must be weaker than the module. The former is taken care of in `typemod.ml` `maybe_infer_modalities`. The latter is take care of by the definition of `apply` for `Diff`.

Reply:
Ah yes, not sure what I was thinking there.